shell/src/main/java/org/apache/kafka/shell/MetadataNodeManager.java

```diff
@@ -403,4 +429,158 @@ static List<String> clientQuotaRecordDirectories(List<EntityData> entityData) {
+                FileNode file = s3StreamObjects.create(Long.toString(record.objectId()));
+                file.setContents(S3StreamObjectRecordJsonConverter.write(record, S3StreamObjectRecord.HIGHEST_SUPPORTED_VERSION).toPrettyString());
+                break;
+            }
+            case REMOVE_S3_STREAM_OBJECT_RECORD: {
+                RemoveS3StreamObjectRecord record = (RemoveS3StreamObjectRecord) message;
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()), "s3StreamObjects", Long.toString(record.objectId()));
+                break;
+            }
+            case S3_STREAM_SET_OBJECT_RECORD: {
+                S3StreamSetObjectRecord record = (S3StreamSetObjectRecord) message;
+                DirectoryNode streamSetObject = data.root.mkdirs("nodes", Integer.toString(record.nodeId()), "s3StreamSetObjects", Long.toString(record.objectId()));
+                streamSetObject.create("orderId").setContents(Long.toString(record.orderId()));
+                streamSetObject.create("dataTimeInMs").setContents(Long.toString(record.dataTimeInMs()));
+                byte[] bytes = record.ranges();
+                // Decode ranges and convert it to human-readable string.
+                List<StreamOffsetRange> ranges = S3StreamSetObject.decode(bytes);
+                streamSetObject.create("ranges").setContents(ranges.toString());
+                break;
+            }
+            case REMOVE_STREAM_SET_OBJECT_RECORD: {
+                RemoveStreamSetObjectRecord record = (RemoveStreamSetObjectRecord) message;
+                data.root.rmrf("nodes", Integer.toString(record.nodeId()), "s3StreamSetObjects", Long.toString(record.objectId()));
+                break;
+            }
+            case S3_OBJECT_RECORD: {
+                S3ObjectRecord record = (S3ObjectRecord) message;
+                DirectoryNode s3Objects = data.root.mkdirs("s3Objects");
+                FileNode file = s3Objects.create(Long.toString(record.objectId()));
+                file.setContents(S3ObjectRecordJsonConverter.write(record, S3ObjectRecord.HIGHEST_SUPPORTED_VERSION).toPrettyString());
+                break;
+            }
+            case REMOVE_S3_OBJECT_RECORD: {
+                RemoveS3ObjectRecord record = (RemoveS3ObjectRecord) message;
+                data.root.rmrf("s3Objects", Long.toString(record.objectId()));
+                break;
+            }
+            case ASSIGNED_STREAM_ID_RECORD: {
+                AssignedStreamIdRecord record = (AssignedStreamIdRecord) message;
+                DirectoryNode streamIdsNode = data.root.mkdirs("s3Streams");
+                // Empty directory here. We just need to create the directory.
+                streamIdsNode.mkdirs(Long.toString(record.assignedStreamId()));
+                break;
+            }
+            case ASSIGNED_S3_OBJECT_ID_RECORD: {
```

Comment:
ASSIGNED_S3_OBJECT_ID_RECORD is the next objectId will be allocated, it may be represent as a file s3object/nextObjectId and the file content is the value.

It's the same for UPDATE_NEXT_NODE_ID_RECORD、ASSIGNED_STREAM_ID_RECORD

Reply:
Nice proposal... If there is an associated `RemoveS3ObjectRecord`, using the folder directory results in the `nextObjectId` being dropped.